contracts/accessless/OlympusChainlinkBasedSwapper.sol

```diff
@@ -0,0 +1,215 @@
+        }
+
+        address denomination = (params.denomination == address(0)) ? Denominations.USD : params.denomination;
+
+        // checks
+        if (params.path.length - 1 != params.fees.length) revert OlympusChainlinkBasedSwapper_WrongV3PathLength();
+
+        // interactions + checks
+
+        bool check = IERC20(inputToken).transferFrom(msg.sender, address(this), params.amount);
+
+        if (!check) revert OlympusChainlinkBasedSwapper_TransferFromFailed();
+
+        check = IERC20(inputToken).approve(address(v3SwapRouter), params.amount);
+
+        if (!check) revert OlympusChainlinkBasedSwapper_ApproveFailed();
+
+        // big interaction
+        if (isIn) {
+            amount = v3SwapRouter.exactInput(
+                ISwapRouter.ExactInputParams({
+                    path: _encode(params.fees.length, params.path, params.fees),
+                    recipient: params.recipient,
+                    deadline: (params.deadline == 0) ? block.timestamp : params.deadline,
+                    amountIn: params.amount,
+                    amountOutMinimum: _calcAmountMin(
+                        inputToken,
+                        outputToken,
+                        denomination,
+                        params.slippage,
+                        params.amount
+                    )
+                })
+            );
+        } else {
+            amount = v3SwapRouter.exactOutput(
+                ISwapRouter.ExactOutputParams({
+                    path: _encode(params.fees.length, params.path, params.fees),
+                    recipient: params.recipient,
+                    deadline: (params.deadline == 0) ? block.timestamp : params.deadline,
+                    amountOut: _calcAmountMin(inputToken, outputToken, denomination, params.slippage, params.amount),
+                    amountInMaximum: params.amount
+                })
+            );
+        }
+
+        // cleanup
+        _clean(inputToken);
+    }
+
+    // @notice
+    function _clean(address inputToken) internal {
+        bool check = IERC20(inputToken).transfer(msg.sender, IERC20(inputToken).balanceOf(address(this)));
+        if (!check) revert OlympusChainlinkBasedSwapper_TransferFailed();
+    }
+
+    function _calcAmountMin(
+        address inputToken,
+        address outputToken,
+        address denom,
+        uint256 slippage,
+        uint256 amount
+    ) internal view returns (uint256 amountMin) {
+        uint256 inPrice = uint256(_latestPrice(inputToken, denom));
+        uint256 outPrice = uint256(_latestPrice(outputToken, denom));
+
+        // slippage must be normalized to 1e18
+        amountMin = (amount * inPrice * slippage) / (1e18 * outPrice);
+        amountMin = (amountMin * IERC20Metadata(outputToken).decimals()) / IERC20Metadata(inputToken).decimals();
+    }
+
+    function _latestPrice(address base, address quote) internal view returns (int256 price) {
+        if (base == weth) base = Denominations.ETH;
```

Comment:
might be cleaner to do `base = base == weth ? Denominations.ETH : Denominations.BTC;`